### DISCUSSION TITLE
non-deterministic hoisted node_modules/.pnpm/node_modules

### DISCUSSION THREAD

**Discussion by rrosenshain-sc** [pnpm/pnpm#7070] [Q&A]
```
pnpm -v
8.7.0
```

I'm using pnpm with a monorepo and seeing weird inconsistencies with how `node_modules/.pnpm/node_modules` are resolved.
I also noticed `modules.yaml` is showing different versions in the 

I don't have a simple repro yet but basically I pull my branch, run `pnpm i`, everything works. I then make a change to package.json manually and then run `pnpm i` again (I think any change to pnpm related files would cause it) and the modules.yaml files is created differently and the node_modules gets hoisted differently.

I'm using the default .npmrc hoisting config
```
// .npmrc
engine-strict = true
enable-pre-post-scripts=true
auto-install-peers=false
```

Some examples of changes in the `modules.yaml` file:
```
hoistPattern:
  - '*'
hoistedDependencies:
...
-   /react-intl/5.17.5(react@16.14.0)(typescript@4.2.4):
+   /react-intl/5.17.5(react@18.2.0)(typescript@5.1.6):
...
-   /react-router-dom/6.15.0(react-dom@16.14.0)(react@16.14.0):
+  /react-router-dom/6.14.2(react-dom@18.2.0)(react@18.2.0):
...
```

Any idea why it might be happening?

**Comment by rrosenshain-sc** [comment]
I have a repro example: https://github.com/swang9-sc/pnpm-demo/tree/main
Basically running pnpm i multiple times generates a different tree randomly

Would been great to at least have a configurable choice somehow - like "lowest/highest" so at least it's deterministic

**Comment by robert-westenberger** [comment]
I'm also seeing this, on pnpm@8.15.1. And i can reproduce it with your demo repo. I don't understand how this isn't considered a critical issue by the pnpm team. in my own repo i am using pnpm i --frozen-lockfile and its still non deterministic.
